pym/bob/cmds/build.py

```diff
@@ -943,3 +943,103 @@ def doClean(argv, bobRoot):
         if not args.dry_run:
             removePath(d)
 
+def doQueryPath(argv, bobRoot):
+    # Local imports
+    from string import Formatter
+
+    # Configure the parser
+    parser = argparse.ArgumentParser(prog="bob query-path",
+                                     description="Lists all packages and their last build directory. If a package has never been built, no output is shown for that package.")
+    parser.add_argument('packages', metavar='PACKAGE', type=str, nargs='+',
+        help="(Sub-)package to build")
+    parser.add_argument('-f', help='Select output format', default='{name}\t{dist}', metavar='FORMAT')
+    parser.add_argument('-D', default=[], action='append', dest="defines",
+        help="Override default environment variable")
+    parser.add_argument('-c', dest="configFile", default=[], action='append',
+        help="Use config File")
+
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('--sandbox', action='store_true', help="Enable sandboxing")
+    group.add_argument('--no-sandbox', action='store_false', dest='sandbox', help="Disable sandboxing")
+    parser.set_defaults(sandbox=None)
+
+    # FIXME: this does not have any effect?
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('--develop', action='store_true',  dest='dev', help="Use developer mode")
+    group.add_argument('--release', action='store_false', dest='dev', help="Use release mode", default=True)
+
+    # Parse args
+    args = parser.parse_args(argv)
+    if args.sandbox == None:
+        args.sandbox = not args.dev
+
+    # Process defines
+    defines = {}
+    for define in args.defines:
+        d = define.split("=")
+        if len(d) == 1:
+            defines[d[0]] = ""
+        elif len(d) == 2:
+            defines[d[0]] = d[1]
+        else:
+            parser.error("Malformed define: "+define)
+
+    # Process the recipes
+    recipes = RecipeSet()
+    recipes.defineHook('releaseNameFormatter', LocalBuilder.releaseNameFormatter)
+    recipes.defineHook('developNameFormatter', LocalBuilder.developNameFormatter)
+    recipes.defineHook('developNamePersister', LocalBuilder.developNamePersister)
+    recipes.setConfigFiles(args.configFile)
+    recipes.parse()
+
+    # Find the name formatter
+    if args.dev:
+        nameFormatter = recipes.getHook('developNameFormatter')
+        developPersister = recipes.getHook('developNamePersister')
+        nameFormatter = developPersister(nameFormatter)
+    else:
+        nameFormatter = recipes.getHook('releaseNameFormatter')
+        nameFormatter = LocalBuilder.releaseNamePersister(nameFormatter)
```

Comment:
Here you should actually use a completely separate function. The `releaseNameFormatter` hook is not needed as we just want to query the persistent state in the release case. The key is to replace LocalBuilder.releaseNamePersister() by something along these lines (not tested):

```python
    @staticmethod
    def releaseNameInterrogator(step, props):
        return BobState().getExistingByNameDirectory(asHexStr(step.getVariantId()))
```

That way the normal path formatting machine should be usable.